shell/platform/fuchsia/flutter/text_delegate.h

```diff
@@ -0,0 +1,152 @@
+      std::unique_ptr<fuchsia::ui::input::InputEvent> event) override;
+
+  /// |fuchsia::ui::input::InputMethodEditorClient|
+  /// Called by the embedder when the action key is pressed, and the requested
+  /// action is supplied to Flutter.
+  void OnAction(fuchsia::ui::input::InputMethodAction action) override;
+
+  /// Gets a new input method editor from the input connection. Run when both
+  /// Scenic has focus and Flutter has requested input with setClient.
+  void ActivateIme();
+
+  /// Detaches the input method editor connection, ending the edit session and
+  /// closing the onscreen keyboard. Call when input is no longer desired,
+  /// either because Scenic says we lost focus or when Flutter no longer has a
+  /// text field focused.
+  void DeactivateIme();
+
+  /// Channel handler for kTextInputChannel
+  bool HandleFlutterTextInputChannelPlatformMessage(
+      std::unique_ptr<flutter::PlatformMessage> message);
+
+  /// Returns true if there is a text state (i.e. if some text editing is in
+  /// progress).
+  bool HasTextState() { return last_text_state_ != nullptr; }
+
+ private:
+  // Activates the input method editor, assigning |action| to the "enter" key.
+  // This action will be reported by |OnAction| above when the "enter" key is
+  // pressed. Note that in the case of multi-line text editors, |OnAction| will
+  // never be called: instead, the text editor will insert a newline into the
+  // edited text.
+  void ActivateIme(fuchsia::ui::input::InputMethodAction action);
+
+  // Converts Fuchsia platform key codes into Flutter key codes.
+  Keyboard keyboard_translator_;
+
+  // A callback for sending a single Flutter platform message.
+  std::function<void(std::unique_ptr<flutter::PlatformMessage>)>
+      dispatch_callback_;
+
+  // TextDelegate server-side binding.  Methods called when the text edit state
+  // is updated.
+  fidl::Binding<fuchsia::ui::input::InputMethodEditorClient> ime_client_;
+
+  // An interface for interacting with a text input control.
+  fuchsia::ui::input::InputMethodEditorPtr ime_;
+
+  // An interface for requesting the InputMethodEditor.
+  fuchsia::ui::input::ImeServicePtr text_sync_service_;
+
+  // The locally-unique identifier of the text input currently in use. Flutter
+  // usually uses only one at a time.
+  int current_text_input_client_ = 0;
+
+  // TextDelegate server side binding. Methods called when a key is pressed.
+  fidl::Binding<fuchsia::ui::input3::KeyboardListener>
+      keyboard_listener_binding_;
+
+  // The client-side stub for calling the Keyboard protocol.
+  fuchsia::ui::input3::KeyboardPtr keyboard_;
+
+  // last_text_state_ is the last state of the text input as reported by the IME
+  // or initialized by Flutter. We set it to null if Flutter doesn't want any
+  // input, since then there is no text input state at all.
+  // If nullptr, then no editing is in progress.
+  std::unique_ptr<fuchsia::ui::input::TextInputState> last_text_state_;
```

Comment:
std::optional might give you the desired behavior expressed more directly. I see you already use std::optional on a different field below. Does this object need to be on the heap?

Reply:
Done.